File: Source/Extensions/Tessa.Extensions.Default.Client/Tiles/KrPermissionsTileExtension.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Windows.Input;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.KrPermissions;
using Tessa.Platform;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;
using Tessa.UI.Tiles.Extensions;

namespace Tessa.Extensions.Default.Client.Tiles
{
    /// <summary>
    /// Расширение на тайл "Сбросить кеш правил доступа".
    /// </summary>
    public sealed class KrPermissionsTileExtension : TileExtension
    {
        #region Fields

        private static readonly ICommand tileCommand = new DelegateCommand(TileAction);

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override Task InitializingGlobal(ITileGlobalExtensionContext context)
        {
            var panel = context.Workspace.LeftPanel;

            panel.Tiles.Add(
                new Tile(
                    DefaultTileNames.KrPermissionsDropCache,
                    TileHelper.SplitCaption("$KrTiles_DropPermissionsCache"),
                    context.Icons.Get("Thin57"),
                    panel,
                    tileCommand,
                    order: 10,
                    evaluating: KrPermissionsTileEvaluating));

            return Task.CompletedTask;
        }

        #endregion

        #region Private Methods

        private static void KrPermissionsTileEvaluating(object sender, TileEvaluationEventArgs e)
        {
            var editor = e.CurrentTile.Context.CardEditor;
            ICardModel model;

            e.SetIsEnabledWithCollapsing(
                e.CurrentTile,
                editor is not null
                && (model = editor.CardModel) is not null
                && model.CardType.ID == DefaultCardTypes.KrPermissionsTypeID);
        }

        private static async void TileAction(object obj)
        {
            var context = UIContext.Current;
            var editor = context.CardEditor;

            if (editor is null)
            {
                return;
            }

            await editor.SaveCardAsync(
                context,
                new Dictionary<string, object>(StringComparer.Ordinal)
                {
                    [KrPermissionsHelper.DropPermissionsCacheMark] = BooleanBoxes.True,
                });
        }

        #endregion
    }
}

File: WebClient SDK/src/default/tiles/krPermissionsTileExtension.ts
import { TileExtension, ITileGlobalExtensionContext, Tile, TileEvaluationEventArgs } from 'tessa/ui/tiles';
import { ICardModel } from 'tessa/ui/cards';
import { UIContext } from 'tessa/ui';
import { systemKeyPrefix } from 'tessa/cards';
import { createTypedField, DotNetType } from 'tessa/platform';

export class KrPermissionsTileExtension extends TileExtension {

  public initializingGlobal(context: ITileGlobalExtensionContext) {
    const panel = context.workspace.leftPanel;

    panel.tiles.push(
      new Tile({
        name: 'KrPermissionsDropCache',
        caption: '$KrTiles_DropPermissionsCache',
        icon: 'icon-thin-057',
        contextSource: panel.contextSource,
        command: this.tileAction,
        order: 10,
        evaluating: this.krPermissionsTileEvaluating
      })
    );
  }

  private krPermissionsTileEvaluating = (e: TileEvaluationEventArgs) => {
    const editor = e.currentTile.context.cardEditor;
    let model: ICardModel;

    e.setIsEnabledWithCollapsing(
      e.currentTile,
      !!editor
      && !!(model = editor.cardModel!)
      && model.cardType.id === 'fa9dbdac-8708-41df-bd72-900f69655dfa'); // KrPermissionsTypeID
  }

  private tileAction = async () => {
    const context = UIContext.current;
    const editor = UIContext.current.cardEditor;

    if (!editor) {
      return;
    }

    await editor.saveCard(
      context,
      {
        [systemKeyPrefix + 'DropPermissionsCache']: createTypedField(true, DotNetType.Boolean)
      }
    );
  }

}